<center>
<img src="../../img/ods_stickers.jpg" />
    
## [mlcourse.ai](https://mlcourse.ai) – Open Machine Learning Course 

Author: [Yury Kashnitsky](https://yorko.github.io). This material is subject to the terms and conditions of the [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license. Free use is permitted for any non-commercial purpose.

# <center> Topic 1. Exploratory data analysis with Pandas
## <center>Practice. Analyzing "Titanic" passengers

**Fill in the missing code ("You code here") and choose answers in a [web-form](https://docs.google.com/forms/d/16EfhpDGPrREry0gfDQdRPjoiQX9IumaL2mPR0rcj19k/edit).**

In [2]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt

# Graphics in SVG format are more sharp and legible
%config InlineBackend.figure_format = 'svg'
pd.set_option("display.precision", 2)

**Read data into a Pandas DataFrame**

In [3]:
data = pd.read_csv("../../data/titanic_train.csv", index_col="PassengerId")

**First 5 rows**

In [4]:
data.head(5)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket   Fare Cabin Embarked  
PassengerId                                                        
1                1      0         A/5 21171   7.25   NaN        S  
2                1      0          PC 17599  71.28   C85        C  
3                0      0  STON/O2. 3101282   7.92   NaN        S  
4                1      0            113803  53.10  C123        S  
5                0      0            373450   8.05   NaN        S

In [5]:
data.describe()

Survived  Pclass     Age   SibSp   Parch    Fare
count    891.00  891.00  714.00  891.00  891.00  891.00
mean       0.38    2.31   29.70    0.52    0.38   32.20
std        0.49    0.84   14.53    1.10    0.81   49.69
min        0.00    1.00    0.42    0.00    0.00    0.00
25%        0.00    2.00   20.12    0.00    0.00    7.91
50%        0.00    3.00   28.00    0.00    0.00   14.45
75%        1.00    3.00   38.00    1.00    0.00   31.00
max        1.00    3.00   80.00    8.00    6.00  512.33

**Let's select those passengers who embarked in Cherbourg (Embarked=C) and paid > 200 pounds for their ticker (fare > 200).**

Make sure you understand how actually this construction works.

In [6]:
data[(data["Embarked"] == "C") & (data.Fare > 200)].head()

Survived  Pclass  \
PassengerId                     
119                 0       1   
259                 1       1   
300                 1       1   
312                 1       1   
378                 0       1   

                                                        Name     Sex   Age  \
PassengerId                                                                  
119                                 Baxter, Mr. Quigg Edmond    male  24.0   
259                                         Ward, Miss. Anna  female  35.0   
300          Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0   
312                               Ryerson, Miss. Emily Borie  female  18.0   
378                                Widener, Mr. Harry Elkins    male  27.0   

             SibSp  Parch    Ticket    Fare            Cabin Embarked  
PassengerId                                                            
119              0      1  PC 17558  247.52          B58 B60        C  
259              0      0  PC 17755  512.33              NaN        C  
300              0      1  PC 17558  247.52          B58 B60        C  
312              2      2  PC 17608  262.38  B57 B59 B63 B66        C  
378              0      2    113503  211.50              C82        C

**We can sort these people by Fare in descending order.**

In [7]:
data[(data["Embarked"] == "C") & (data["Fare"] > 200)].sort_values(
    by="Fare", ascending=False
).head()

Survived  Pclass                                   Name     Sex  \
PassengerId                                                                    
259                 1       1                       Ward, Miss. Anna  female   
680                 1       1     Cardeza, Mr. Thomas Drake Martinez    male   
738                 1       1                 Lesurer, Mr. Gustave J    male   
312                 1       1             Ryerson, Miss. Emily Borie  female   
743                 1       1  Ryerson, Miss. Susan Parker "Suzette"  female   

              Age  SibSp  Parch    Ticket    Fare            Cabin Embarked  
PassengerId                                                                  
259          35.0      0      0  PC 17755  512.33              NaN        C  
680          36.0      0      1  PC 17755  512.33      B51 B53 B55        C  
738          35.0      0      0  PC 17755  512.33             B101        C  
312          18.0      2      2  PC 17608  262.38  B57 B59 B63 B66        C  
743          21.0      2      2  PC 17608  262.38  B57 B59 B63 B66        C

**Let's create a new feature.**

In [8]:
def age_category(age):
    """
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    """
    if age < 30:
        return 1
    elif age < 55:
        return 2
    elif age >= 55:
        return 3

In [9]:
age_categories = [age_category(age) for age in data.Age]
data["Age_category"] = age_categories

**Another way is to do it with `apply`.**

In [10]:
data["Age_category"] = data["Age"].apply(age_category)

**1. How many men/women were there onboard?**
- 412 men and 479 women
- 314 men and 577 women
- 479 men and 412 women
- 577 men and 314 women ++

In [11]:
# You code here
data["Sex"].value_counts()

male      577
female    314
Name: Sex, dtype: int64

In [12]:
(data["Sex"] == "male").sum(),  (data["Sex"] == "female").sum()

(577, 314)

**2. Print the distribution of the `Pclass` feature. Then the same, but for men and women separately. How many men from second class were there onboard?**
- 104
- 108
- 112
- 125

In [13]:
# You code here

pd.crosstab(data["Sex"], data["Pclass"], margins=True)

Pclass    1    2    3  All
Sex                       
female   94   76  144  314
male    122  108  347  577
All     216  184  491  891

**3. What are median and standard deviation of `Fare`?. Round to two decimals.**
- +median is  14.45, standard deviation is 49.69
- median is 15.1, standard deviation is 12.15
- median is 13.15, standard deviation is 35.3
- median is  17.43, standard deviation is 39.1

In [14]:
# You code here
data["Fare"].round(2).median(), round(data["Fare"].std(), 2)

(14.45, 49.69)

**4. Is that true that the mean age of survived people is higher than that of passengers who eventually died?**
- Yes
- +No+


In [15]:
# You code here
data.groupby("Survived")["Age"].agg("mean")


Survived
0    30.63
1    28.34
Name: Age, dtype: float64

**5. Is that true that passengers younger than 30 y.o. survived more frequently than those older than 60 y.o.? What are shares of survived people among young and old people?**
- 22.7% among young and 40.6% among old
- +40.6% among young and 22.7% among old+
- 35.3% among young and 27.4% among old
- 27.4% among young and  35.3% among old

In [16]:
young_people = data.loc[data["Age"] < 30.0, "Survived"]
old_people = data.loc[data["Age"] > 60.0, "Survived"]


In [17]:
round(young_people.value_counts(normalize=True) * 100, 1)

0    59.4
1    40.6
Name: Survived, dtype: float64

In [18]:
round(old_people.value_counts(normalize=True) * 100, 1)

0    77.3
1    22.7
Name: Survived, dtype: float64

**6. Is that true that women survived more frequently than men? What are shares of survived people among men and women?**
- 30.2% among men and 46.2% among women
- 35.7% among men and 74.2% among women
- 21.1% among men and 46.2% among women
- +18.9% among men and 74.2% among women+

In [19]:
# You code here
man_survived = data.loc[data["Sex"] == "male", "Survived"]
woman_survived = data.loc[data["Sex"] == "female", "Survived"]

In [20]:
print("Shares of survived people: \n\n amound of men: \n {} \n amound of women: \n {}"
.format(round(man_survived.value_counts(normalize=True) * 100, 1), (round(woman_survived.value_counts(normalize=True) * 100, 1))))

Shares of survived people: 

 amound of men: 
 0    81.1
1    18.9
Name: Survived, dtype: float64 
 amound of women: 
 1    74.2
0    25.8
Name: Survived, dtype: float64


**7. What's the most popular first name among male passengers?**
- Charles
- Thomas
- <font color = "Green">William</font>
- John

In [21]:
data["Name"][1]

'Braund, Mr. Owen Harris'

In [22]:
data.query("Sex == 'male'")

Survived  Pclass                            Name   Sex   Age  \
PassengerId                                                                 
1                   0       3         Braund, Mr. Owen Harris  male  22.0   
5                   0       3        Allen, Mr. William Henry  male  35.0   
6                   0       3                Moran, Mr. James  male   NaN   
7                   0       1         McCarthy, Mr. Timothy J  male  54.0   
8                   0       3  Palsson, Master. Gosta Leonard  male   2.0   
...               ...     ...                             ...   ...   ...   
884                 0       2   Banfield, Mr. Frederick James  male  28.0   
885                 0       3          Sutehall, Mr. Henry Jr  male  25.0   
887                 0       2           Montvila, Rev. Juozas  male  27.0   
890                 1       1           Behr, Mr. Karl Howell  male  26.0   
891                 0       3             Dooley, Mr. Patrick  male  32.0   

             SibSp  Parch            Ticket   Fare Cabin Embarked  \
PassengerId                                                         
1                1      0         A/5 21171   7.25   NaN        S   
5                0      0            373450   8.05   NaN        S   
6                0      0            330877   8.46   NaN        Q   
7                0      0             17463  51.86   E46        S   
8                3      1            349909  21.07   NaN        S   
...            ...    ...               ...    ...   ...      ...   
884              0      0  C.A./SOTON 34068  10.50   NaN        S   
885              0      0   SOTON/OQ 392076   7.05   NaN        S   
887              0      0            211536  13.00   NaN        S   
890              0      0            111369  30.00  C148        C   
891              0      0            370376   7.75   NaN        Q   

             Age_category  
PassengerId                
1                     1.0  
5                     2.0  
6                     NaN  
7                     2.0  
8                     1.0  
...                   ...  
884                   1.0  
885                   1.0  
887                   1.0  
890                   1.0  
891                   2.0  

[577 rows x 12 columns]

In [23]:
data.loc[1, "Name"].split(".")[1].split()[0]

'Owen'

In [24]:
new_df = data[data["Sex"].apply(lambda sex: sex[:] == "male")]
# for i, dt in new_df.Name.items():
#     new_df["First_name"] = data.loc[i,"Name"].split(".")[1].split()[0]

In [25]:
First_name = data.loc[data["Sex"] == "male", "Name"].apply(lambda name: name.split(".")[1].split()[0])
First_name.value_counts()

William    35
John       25
George     14
Charles    13
Thomas     13
           ..
Ignjac      1
Yoto        1
Austen      1
Mitto       1
Juozas      1
Name: Name, Length: 287, dtype: int64

**8. How is average age for men/women dependent on `Pclass`? Choose all correct statements:**
- <font color = "green">On average, men of 1 class are older than 40</font>
- On average, women of 1 class are older than 40
- <font color = "green">Men of all classes are on average older than women of the same class </font>
- <font color = "green"> On average, passengers of the first class are older than those of the 2nd class who are older than passengers of the 3rd class </font>

In [31]:
# You code here
# data.oc[data["Sex"] == "male", "Pclass"].value_counts()

data.groupby(["Sex", "Pclass"], as_index=False).agg({"Age": "mean"})

Sex  Pclass    Age
0  female       1  34.61
1  female       2  28.72
2  female       3  21.75
3    male       1  41.28
4    male       2  30.74
5    male       3  26.51

More nicely

In [39]:
pd.crosstab(data["Sex"], data["Pclass"], values=data["Age"], aggfunc=np.mean)

Pclass      1      2      3
Sex                        
female  34.61  28.72  21.75
male    41.28  30.74  26.51

## Useful resources
* The same notebook as an interactive web-based [Kaggle Kernel](https://www.kaggle.com/kashnitsky/topic-1-practice-analyzing-titanic-passengers) with a [solution](https://www.kaggle.com/kashnitsky/topic-1-practice-solution)
* Topic 1 "Exploratory Data Analysis with Pandas" as a [Kaggle Kernel](https://www.kaggle.com/kashnitsky/topic-1-exploratory-data-analysis-with-pandas)
* Main course [site](https://mlcourse.ai), [course repo](https://github.com/Yorko/mlcourse.ai), and YouTube [channel](https://www.youtube.com/watch?v=QKTuw4PNOsU&list=PLVlY_7IJCMJeRfZ68eVfEcu-UcN9BbwiX)